### ISSUE TITLE
Handling of default values

### ISSUE THREAD

**Issue by felixfbecker** [tintoy/dotnet-kube-client#28]
I got diffing objects working well so far, but there is one problem: Several properties in Kubernetes objects are scalars with default values. For example:
- `int ProbeV1.FailureThreshold` defaults to `3`
- `string ServicePortV1.Protocol` defaults to `"TCP"`
and many more.
These properties are often not declared in YAML configuration.
But when deserializing the YAML, the properties get the C# type default value, instead of the server default value. I.e. `FailureThreshold` is always `0`, `Protocol` is always `null`.
When diffing against the server, this will then generate a patch that sets these values to the C# default value (`0`/`null`).

How should we handle this?

- When diffing, check if the new value is the C# default value, and exclude it from the patch. This would make it impossible to actually set a field to `0` or `null`. Does the API have fields anywhere that can be set to the C#/Go default value? E.g. for `FailureThreshold` the minimum is `1`, and in other places where `0` is valid I saw them use nullable/pointer types, e.g. `ConfigMapVolumeSource.DefaultMode`. But there are a lot of booleans that are not nullable, and this would make it impossible to switch a boolean back to `false` after it was set to `true` on the server.
- Apply the server defaults client side. This would break if they ever change the default. The defaults are only exposed through documentation, not in the schema.
- Make everything nullable. This would be a bad experience when querying _from_ the server as these are always defined.
- Use Dictionaries or `dynamic` instead of strongly typed objects for deserialisation. I would like to avoid this because then we lose the formatting and autocompletion in PowerShell, and it becomes way harder to look up the merge strategies to use for properties...

**Comment by tintoy** [comment]
I like your idea - let's do that.

**Comment by felixfbecker** [comment]
So even if we have `DefaultValue` attributes on everything and initialisers, it would still mean that if a YAML does not declare a field and the value is different than the default on the server, it will reset the value to the default value. I wonder if that is acceptable behaviour or not.
If we always ignore default values, it would be impossible to reset a value back to the default value.

Maybe we _do_ need to make everything with a default nullable? So that `null` means "not specified"/"ignore" while the actual default value means "reset". The OpenAPI spec is inconsistent about this too - some scalars are nullable _in addition_ to having a default value, while others are not.

**Comment by felixfbecker** [comment]
Maybe we need to generate two sets of classes for API responses and for inputs? Where inputs just has more optional fields, so the API responses can extend the input classes and override the fields with default values to make them non-optional.

What do you usually do in C# in such cases? In TypeScript I can distinguish between `null` and `undefined` and scalars don't have zero values (`undefined` is the zero value for everything).

**Comment by tintoy** [comment]
Hmm, not sure about yamldotnet but JSON.NET has DefaultValueHandling so if the value is some well-known scalar value (or null) the member won't be serialised. Perhaps yamldotnet has something similar?

**Comment by tintoy** [comment]
https://github.com/aaubry/YamlDotNet/issues/21

**Comment by felixfbecker** [comment]
Serialisation is not the problem, it's generating the patch (diffing yaml vs server state)

**Comment by tintoy** [comment]
Ah. Yeah I could apply the DefaultValue attribute and you could look for that?

**Comment by tintoy** [comment]
You could treat that default value as "missing" or "unspecified" perhaps?

**Comment by tintoy** [comment]
If the attribute is.missing th

**Comment by tintoy** [comment]
Man, I hate doing GitHub on my phone :)

**Comment by tintoy** [comment]
If attribute is missing there is a way to determine default value for type but I'd have to look it up.

**Comment by tintoy** [comment]
Here you go:

https://stackoverflow.com/a/353073/885866

**Comment by felixfbecker** [comment]
Yeah, but if you always ignore the default value, then it would become impossible to set the field to the default value again after it was modified on the server...

**Comment by felixfbecker** [comment]
Here's another solution I just came up with:

Generate setters for every field (and backing properties) that will mark the property as dirty when set in a HashSet that can be queried publicly. E.g.
```csharp
class KubeModel {
	protected HashSet<string> dirtyProperties = new HashSet<string>();
	public IImmutableSet<string> DirtyProperties {
		get { return dirtyProperties; }
	}
}
class DeploymentSpec extends KubeModel {
 	private int progressDeadlineSeconds;
    public int ProgressDeadlineSeconds {
   		get { return progressDeadlineSeconds; }
 		set {
  			progressDeadlineSeconds = value; 	
			dirtyProperties.Add("ProgressDeadlineSeconds");
		}
    }
}
```
It would be more code but it's generated, so doesn't really matter.
This would not require changing types to nullable and also remove the need for maintaining a map of default values and even non-updateable properties (#29), because those properties would never get set by the YAML deserialisation.
The diff implementation would ignore any properties that are not dirty.

The only other solution I can think of is to deserialise to Dictionaries, then build `PSObject`s and return those, but "lie" about the `OutputType` in the Cmdlet to make autocompletion still work. While building `PSObject`s, the "virtual" type needs to be added as a TypeName to make output formatting work.
Then make the patch cmdlet convert back to dictionaries first, and make the diff implementation "track" the "virtual" model type of the dictionaries by starting at `KubeResource` and traversing through the `Type` as its traversing through the dictionaries, so it can retrieve the patch strategies etc from attributes.
The disadvantage is that strictly `Get-Kube*` cmdlets now return a completely different type than `ConvertFrom-KubeYaml`.

What do you think?

**Comment by felixfbecker** [comment]
What do you mean with as-is? Dictionary? We need to give it _some_ type to serialize into

**Comment by tintoy** [comment]
What if we generate a second set of models in parallel to the first (in a namespace like `KubeClient.Models.Tracked`) which have the tracking behaviour and implicit / explicit cast operators (or `AsTracked()` / `AsNonTracked()`) to convert back and forth? The actual code for tracking changes can be improved using the `[CallerMemberName]` attribute (or whatever it's called).

**Comment by tintoy** [comment]
Not for the purposes of diff perhaps? Can't remember for yamldotnet but JSON.NET has JObject / Jtoken (a DOM for JSON, effectively). I imagine yamldotnet must have something similar? So I mean don't deserialise at all if you don't need to...

**Comment by tintoy** [comment]
Yep, they have a DOM for YAML:

https://github.com/aaubry/YamlDotNet/wiki/Samples.LoadingAYamlStream

**Comment by felixfbecker** [comment]
Yeah, but we can't operate on the YAML string ;)
I think you mean what I wrote in my second idea:

> The only other solution I can think of is to deserialise to Dictionaries, then build PSObjects and return those, but "lie" about the OutputType in the Cmdlet to make autocompletion still work. While building PSObjects, the "virtual" type needs to be added as a TypeName to make output formatting work.
> Then make the patch cmdlet convert back to dictionaries first, and make the diff implementation "track" the "virtual" model type of the dictionaries by starting at KubeResource and traversing through the Type as its traversing through the dictionaries, so it can retrieve the patch strategies etc from attributes.
> The disadvantage is that strictly Get-Kube* cmdlets now return a completely different type than ConvertFrom-KubeYaml.

If we go with separate class hierarchies, we'll have the same problem of fragmented types...

**Comment by felixfbecker** [comment]
> Yep, they have a DOM for YAML

I believe the patch implementation shouldn't rely on the encoding being YAML though, so Dictionaries would be the better choice.

**Comment by tintoy** [comment]
> If we go with separate class hierarchies, we'll have the same problem of fragmented types

A fair point.

We could, however, change the client so that its methods take / return a `JObject`/`JToken` 
(or a `YamlDocument`) instead of specific model types (i.e. _no_ serialisation / deserialisation at all). Clients can then just use `JObject.ToObject<TModel>()`/`JObject.ToObject(modelType)` _if required_.

**Comment by tintoy** [comment]
I'm pretty comfortable with `JToken` and friends being passed around (it's a reasonably common way to pass untyped or semi-typed data around when that data was received in JSON format in the first place but the schema is not fixed); there's even reasonably high-fidelity conversion between `YamlDocument` and JSON:

https://github.com/aaubry/YamlDotNet/wiki/Samples.ConvertYamlToJson

**Comment by tintoy** [comment]
Seen this? https://github.com/wbish/jsondiffpatch.net

**Comment by felixfbecker** [comment]
I wouldn't want to have PowerShell users needing to do that though...

I have this vision of cmdlets that can all take the same strongly typed Kubernetes model objects as input and all output the same strongly typed objects. Trying to find a solution that gets as close to that as possible 🤔 

> Seen this? https://github.com/wbish/jsondiffpatch.net

Yeah I've seen that, it's not JSON Patch (RFC 6902) though. Kubernetes API only supports RFC 6902 JSON Patch, RFC 7386 JSON Merge Patch and their custom Strategic Merge Patch. I therefor went with JSON Patch.

**Comment by tintoy** [comment]
I agree that's the ideal way to go - but perhaps we might be able to simplify things a bit by keeping the 2 scenarios separate?

1. Consumer wants to load objects, modify them, and then patch.
2. Consumer wants to apply Kube YAML from a file.

For 1, this is a moot point because all fields are already populated; by definition, only fields they change will be changed (and need patch operations generated for them).

For 2, we have a set of keys from the YAML so we know what patch operations are required.

Have I missed anything, or does that more or less cover both scenarios?

**Comment by tintoy** [comment]
(so for 1, just do a property-level diff; all properties will have been populated before modifying)

**Comment by tintoy** [comment]
(as for 2, I know JSON.NET has an operation called `Populate` but I don't know if YamlDotNet does - populate simply deserialises on top of an existing object instance, only setting members for which serialised properties are present in the JSON - it's super useful)

**Comment by tintoy** [comment]
It's also possible to wrap a given model instance in a `PSObject` and augment / replace members that way (e.g. by adding change-tracking behaviour) although this may not be the simplest option...

**Comment by tintoy** [comment]
What if `ConvertFrom-KubeYaml` wrapped each deserialised resource model in a `PSObject` with an added `PSNoteProperty` called something like `YamlKeys` or `LoadedYamlMembers`? That way, you know which values were present in the Yaml to use as a hint :-)

Mind you, if they load from YAML and then modify the resources before passing to `Update-KubeResource` then that might not help so much :-(

**Comment by tintoy** [comment]
(with the option above, the consumer always still gets real, strongly-typed models; it's just that `ConvertFrom-Yaml` adds extra metadata to those deserialised instances to help guide `Update-KubeResource`)

**Comment by felixfbecker** [comment]
One big use case I have is actually loading YAML into an object, mutating it (in PowerShell), then applying it (sending a patch to the API).

For example this script in CI:
- build Docker images
- push them to a registry
- tag them with the currently built commit
- read all Kubernetes YAML (in the repo)  
  `Get-ChildItem -Recurse -File *.yml | ConvertFrom-KubeYaml`
- for all Deployments, modify the `image` fields in the PodSpecs to point to the tag of the currently built commit
  `| ForEach-Object { if ($_ is [DeploymentV1Beta]) { $_.Spec.Containers.Image = $commit } }`
- apply all patches
  `| Update-KubeResource`

For writing this script, it would be tremendously helpful if I had IntelliSense on `$_.Spec.Containers.Image` (i.e. it's strongly typed).

I couldn't find good examples of how to use `PSObject` in C# to add the note property and also read it again, while still declaring types for parameters and output...

**Comment by tintoy** [comment]
Here's a quick example (from memory; pretty sure it'll compile but not _100%_ certain):

```csharp
class MyModel
{
    public string Foo { get; set; }
}

class MyCmdlet
{
    public override void ProcessRecord()
    {
        MyModel model = new MyModel
        {
            Foo = "bar"
        };

        PSObject result = new PSObject(model);
        result.Properties.Add(
            new PSNoteProperty("Name", "MyModel1")
        );

        WriteObject(result);
    }
}
```

**Comment by tintoy** [comment]
Intellisense will work on the result of the Cmdlet because it really is a `MyModel` instance, just wrapped in a `PSObject` (in fact, from memory _all_ objects are wrapped in `PSObject` under the covers but I might be misremembering that).

Additionally, if you decorate your Cmdlet class with`[OutputType(typeof(MyModel))]` or `[OutputType(typeof(MyModel), ParameterSet="MyParameterSet")]` then consumers will get intellisense before it's even run (e.g. when writing a pipeline).

**Comment by tintoy** [comment]
I have an idea for attaching properties to arbitrary objects (CLR-level, not PS Object-level) - give me an hour or so and I'll post some code :)

**Comment by tintoy** [comment]
Ok, forget that but how about this:

When deserialising from YAML, clone the deserialised model and store it as a `PSNoteProperty` (called "Original") on the `PSObject`-wrapped model written to the pipeline. That way you can compare the model's properties with the original to see which ones have been changed! Not certain but you might even be able to mark that property as hidden...

**Comment by tintoy** [comment]
Can do this in C#:
 http://ramblingcookiemonster.github.io/Decorating-Objects/

**Comment by felixfbecker** [comment]
Yeah I think adding the noteproperty would work, but how would you read it from a cmdlet that expects `KubeResouce` as a parameter?

**Comment by tintoy** [comment]
Bah, didn't think of that :)

Maybe have it take a PSObject and use a custom validation attribute to verify the type?

**Comment by tintoy** [comment]
Maybe add a parameter set that takes a different parameter of type PSObject from the pipeline?

**Comment by tintoy** [comment]
You can still get at the wrapped model via the `BaseObject` property.

**Comment by tintoy** [comment]
So just to make sure I understand how this works:

1. Get resource spec from YAML.
2. Consumer may modify that resource spec. 
3. Get resource spec from K8s.
4. Decide what to do?

What if when we loaded the resource spec from YAML we deserialised it on top of the spec loaded from K8s? Then comparing the 2 might be easier...

**Comment by tintoy** [comment]
(all props will be filled in from K8s and they make their changes to that model and when updating K8s all we need to do is load from K8s and compare).

Maybe if `ConvertFrom-KubeYaml` had a switch to enable filling in of undeclared properties from current K8s state?)

**Comment by felixfbecker** [comment]
Like this?
```powershell
Get-Content -Raw deploy.yml | ConvertFrom-KubeYaml -OnTopOfServerState
```
So that would require `ConvertFrom-KubeYaml` to have API access, and essentially would take the job of diffing and patching, because to deserialise "on top of" the server object it has to consider all the aspects of merge strategies etc. So really it just means merging cmdlets into one. It would be nice if we could keep some SoC

**Comment by tintoy** [comment]
SoC?

**Comment by tintoy** [comment]
How about a separate Cmdlet to do merge into model from K8s state?

**Comment by felixfbecker** [comment]
Sorry, separation of concerns. I.e. not have one cmdlet that is solely responsible for parsing, diffing and querying the server. Because there may be use cases that we are not thinking of where one wants to intercept these steps.

**Comment by felixfbecker** [comment]
> How about a separate Cmdlet to do merge into model from K8s state?

That's `Compare-KubeResource`. This whole problem is basically "what type should `Compare-KubeResource` take as a parameter".

**Comment by tintoy** [comment]
`Update-KubeModel` perhaps?

**Comment by tintoy** [comment]
Ergonomics would suffer if they had to explicitly load from YAML, load from K8s, compare, and update. But if the was another Cmdlet or function that wrapped up using the 3 operations together...

**Comment by tintoy** [comment]
That way you cater to the most common use case but preserve composability.

**Comment by tintoy** [comment]
Ah, I was thinking of segregation of responsibility but yeah now I get you.

**Comment by felixfbecker** [comment]
Yes, my plan is to have "shortcuts" that compose these together, but still have separate cmdlets under the hood.

`Update-KubeModel` already exists as `(Compare-KubeResource).ApplyTo()`. But that doesn't solve the problem, because there still needs to be a well-typed way to output the data from `ConvertFrom-KubeYaml` to `Compare-KubeResource`.

**Comment by tintoy** [comment]
I think where I am coming from is that if the model they go to modify was already populated from K8s and then yaml then you don't need to care which properties they changed - just diff and go :)

**Comment by felixfbecker** [comment]
Yes, but unless I'm missing something to "populate it from K8S" you need to diff it

**Comment by tintoy** [comment]
Maybe not? Simply replace all property values with what is specified in the yaml. Then diff and merge only happens when you want to apply...

**Comment by felixfbecker** [comment]
If you only do that shallow (for top-level properties) then you will e.g. override the whole `DeploymentSpec` for a `Deployment`. So you would still have all the problems for nested objects. To do it deep, you need to traverse the object, which can contain lists, like `PodSpec.Containers`. To merge those, you need to consider the merge strategy, etc. So it's basically the full patch algorithm.

**Comment by felixfbecker** [comment]
I.e. after you did that, there is no point in then diffing that object _again_ to generate the patch and send it to the server, you could have just generated a patch in the first place instead of merging the objects on the client.

**Comment by tintoy** [comment]
Oh, I was just thinking in terms of SoC. The later Cmdlets still have to deal with merge but at least the load / consumer-modifications steps don't have to care. But perhaps I'm just misunderstanding the problem (it's been a long day :grin:)

**Comment by tintoy** [comment]
@felixfbecker - ok, had a think about it overnight and this will be the easiest way to attach arbitrary properties (e.g. Original) to models without having to use `PSObject`:

```csharp
using System;
using System.Collections.Generic;
using System.Runtime.CompilerServices;

namespace AttachArbitraryProperties
{
    static class Program
    {
        static void Main()
        {
            Model current;

            current.SetOriginal(
                Clone(current) // e.g. serialise and deserialise, or deserialise twice
            );

            // Later:

            Model original = current.GetOriginal();

            // TODO: Diff
        }
    }

    abstract class Model
    {
        // Whatever
    }

    static class ModelCustomProperties
    {
        static readonly ConditionalWeakTable<object, object> _originals = new ConditionalWeakTable<object, object>();

        public static object GetOriginal(this Model model)
        {
            if (model == null)
                throw new ArgumentNullException(nameof(model));

            if (_originals.TryGetValue(model, out object original))
                return original;

            return null;
        }

        public static void SetOriginal(this Model model, object original)
        {
            if (model == null)
                throw new ArgumentNullException(nameof(model));

            _originals.AddOrUpdate(model, original);
        }
    }
}
```

Will work as long as it's in the same AppDomain, and the `ConditionalWeakTable` won't prevent models from being garbage-collected if nothing else is holding on to them.

Docs here:

https://docs.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.conditionalweaktable-2?view=netcore-2.1

**Comment by felixfbecker** [comment]
Hmm, that'll also not work over remote sessions, right?

**Comment by tintoy** [comment]
No, but it seems (to me at least) to be a sufficiently edge-case scenario that docs may be good enough to indicate that it isn't supported if you load the YAML on a remote system but do the diff on a local system. And you could always add a Cmdlet (e.g. `Update-KubeModelSnapshot`) that could do the augmentation locally before you do any modifications (with docs indicating that this is unnecessary unless you're doing load remotely but compare locally).

BTW, have you seen [platyPS](https://github.com/PowerShell/platyPS)? Pretty easy way to author help for binary Powershell modules :)

**Comment by felixfbecker** [comment]
I just thought about the approach of saving the raw YAML dictionaries in a PSNoteProperty again, but I realised that won't work - the user may add a field dynamically that was not in the original YAML, but then the new field would still be ignored (unless the user also updates the dictionary, which would not be good UX).

What if we took a look at the separate `Models.Tracked` namespace approach? I think this could work - for parsing from YAML, I would always use that, when querying from the server, it's okay if the result is not tracked, because those objects always have all default values filled out. I would propose we make the tracked models all inherit from the non-tracked models.

**Comment by tintoy** [comment]
Ok yeah can do - will make properties virtual and override them on tracked models.

**Comment by tintoy** [comment]
Do you need collections tracked too?

**Comment by felixfbecker** [comment]
No, only model properties.

**Comment by tintoy** [comment]
Ok, easy enough :)

I can get that done sometime on Saturday - will post here when it's ready to try out.

**Comment by tintoy** [comment]
Each tracked model will have a hashset of field names that have been modified. It will be reset during serialisation. Ok?

**Comment by tintoy** [comment]
Sorry, *de*serialisation

**Comment by felixfbecker** [comment]
The hashset could also never be reset for the lifetime of the object.

**Comment by tintoy** [comment]
Ok, made a first attempt at this:

https://github.com/tintoy/dotnet-kube-client/tree/feature/tracked-models

CI build in progress so you can try out the packages.

https://travis-ci.org/tintoy/dotnet-kube-client/builds/422831113

Each model in `KubeClient.Models.Tracked` (providing it has any non-collection properties) called `__ModifiedProperties__`. This is a `HashSet<string>` containing the CLR name of any properties that were modified. I was going to use the JSON / YAML property name but that makes reflection harder than it needs to be.

**Comment by tintoy** [comment]
For example:

https://github.com/tintoy/dotnet-kube-client/blob/feature/tracked-models/src/KubeClient/Models/Tracked/generated/PodV1.cs

**Comment by tintoy** [comment]
Still trying to decide whether to move the tracked models to a separate assembly; this would improve ergonomics for existing users since they won't be confused by 2 models with the same name in different namespaces.

**Comment by felixfbecker** [comment]
Awesome! What's missing it seems is a way to get the right tracked model type like I can for non-tracked through `ModelMetadata`

**Comment by felixfbecker** [comment]
I added that myself on the feature branch.
What I couldn't figure out how to fix is that tracked models are not tracked "deep", i.e. `KubeClient.Models.Tracked.DeploymentV1Beta1#Spec` is of type `KubeClient.Models.DeploymentSpecV1Beta1`, not `KubeClient.Models.Tracked.DeploymentSpecV1Beta1` as it would need to be.

**Comment by tintoy** [comment]
I can fix that - will have a go sometime today :)

**Comment by felixfbecker** [comment]
I'll also do a try of serialising into `dynamic` and see how that approach plays out.

**Comment by felixfbecker** [comment]
I implemented serialising into `PSObject` and diffing `PSObject` and it works well. I am setting type names so output formatting is preserved. I think this is the best approach, so we can abandon the tracked models.

**Comment by tintoy** [comment]
@felixfbecker I think this may have been dealt with as part of the work for tintoy/dotnet-kube-client#43. If you get any spare time, can you have a look and let me know if it works for you?
